PolymerTemplate parsing is not thread safe in production mode

Prerequisite: 
* run in production mode

If two threads enter https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java#L120 simultaneously when the template is not yet cached, the end result is unpredictable.

Both threads will get `useCache=false` and a reference to the same `ParserData` instance. Both threads will then proceed to parse the template and this will fail at some point because of race conditions as nothing inside `ParserData` is synchronized or otherwise designed to be thread safe.

In Bakery tests, you can see this as failures to map fields in templates, e.g.
```
Caused by: java.lang.IllegalStateException: There is no element with id='search' in the template file. Cannot map it using @Id
```
This happens only right after you start the server as the cache is later on populated and used correctly.

The problems appear to be:
* You can't make a cache so that you first check if there is a value, and if not, you populate it with a new value. The operation should be atomic, i.e. `ReflectionCache.contains` should probably not exist at all
* Populating the cache value must happen only once. With the current architecture it would have to happen at https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java#L65
* A `TemplateInitializer` is constructed with a template and a parser. The given parser is not used as a cached version of the parsed template can be used instead. This makes little sense.

Looks like the whole architecture of the class needs to be refactored

Cc @alexberazouski

--- a/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/DefaultTemplateParser.java
+++ b/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/DefaultTemplateParser.java
@@ -22,8 +22,6 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
 import java.util.concurrent.atomic.AtomicBoolean;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 import java.util.stream.Collectors;
 
 import javax.servlet.ServletContext;
@@ -33,17 +31,19 @@ import org.jsoup.nodes.Comment;
 import org.jsoup.nodes.Document;
 import org.jsoup.nodes.Element;
 import org.jsoup.nodes.Node;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import com.vaadin.flow.component.dependency.HtmlImport;
 import com.vaadin.flow.internal.AnnotationReader;
 import com.vaadin.flow.internal.ReflectionCache;
 import com.vaadin.flow.server.DependencyFilter;
+import com.vaadin.flow.server.DependencyFilter.FilterContext;
 import com.vaadin.flow.server.VaadinRequest;
 import com.vaadin.flow.server.VaadinService;
 import com.vaadin.flow.server.VaadinSession;
 import com.vaadin.flow.server.VaadinUriResolverFactory;
 import com.vaadin.flow.server.WrappedHttpSession;
-import com.vaadin.flow.server.DependencyFilter.FilterContext;
 import com.vaadin.flow.shared.ui.Dependency;
 import com.vaadin.flow.shared.ui.Dependency.Type;
 
@@ -53,15 +53,28 @@ import com.vaadin.flow.shared.ui.Dependency.Type;
  * The implementation scans all HTML imports annotations for the given template
  * class and tries to find the one that contains template definition using the
  * tag name.
+ * <p>
+ * The class is Singleton. Use {@link DefaultTemplateParser#getInstance()} to
+ * get its instance.
  *
  * @see TemplateParser
  *
  * @author Vaadin Ltd
  *
  */
-public class DefaultTemplateParser implements TemplateParser {
+public final class DefaultTemplateParser implements TemplateParser {
     private static final ReflectionCache<PolymerTemplate<?>, AtomicBoolean> LOG_CACHE = new ReflectionCache<>(
             clazz -> new AtomicBoolean());
+
+    private static final TemplateParser INSTANCE = new DefaultTemplateParser();
+
+    private DefaultTemplateParser() {
+        // Doesn't allow external instantiation
+    }
+
+    public static TemplateParser getInstance() {
+        return INSTANCE;
+    }
 
     @Override
     public Element getTemplateContent(Class<? extends PolymerTemplate<?>> clazz,
@@ -98,7 +111,8 @@ public class DefaultTemplateParser implements TemplateParser {
             String path = resolvePath(request, url);
 
             if (logEnabled) {
-                getLogger().debug("Html import path '{}' is resolved to '{}'", url, path);
+                getLogger().debug("Html import path '{}' is resolved to '{}'",
+                        url, path);
             }
             try (InputStream content = context.getResourceAsStream(path)) {
                 if (content == null) {
@@ -108,8 +122,10 @@ public class DefaultTemplateParser implements TemplateParser {
                 }
                 Element templateElement = parseHtmlImport(content, url, tag);
                 if (logEnabled && templateElement != null) {
-                    getLogger().debug("Found a template file containing template "
-                        + "definition for the tag '{}' by the path '{}'", tag, url);
+                    getLogger().debug(
+                            "Found a template file containing template "
+                                    + "definition for the tag '{}' by the path '{}'",
+                            tag, url);
                 }
 
                 if (templateElement != null) {
@@ -119,7 +135,8 @@ public class DefaultTemplateParser implements TemplateParser {
             } catch (IOException exception) {
                 // ignore exception on close()
                 if (logEnabled) {
-                    getLogger().warn("Couldn't close template input stream", exception);
+                    getLogger().warn("Couldn't close template input stream",
+                            exception);
                 }
             }
         }

--- a/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/PolymerTemplate.java
+++ b/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/PolymerTemplate.java
@@ -80,7 +80,7 @@ public abstract class PolymerTemplate<M extends TemplateModel>
      * functionality.
      */
     public PolymerTemplate() {
-        this(new DefaultTemplateParser());
+        this(DefaultTemplateParser.getInstance());
     }
 
     /**

--- a/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java
+++ b/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java
@@ -85,7 +85,7 @@ public class TemplateInitializer {
     }
 
     private interface InjectableFieldCunsumer {
-        public void apply(Field field, String id, String tag);
+        void apply(Field field, String id, String tag);
     }
 
     /**
@@ -419,8 +419,7 @@ public class TemplateInitializer {
     /* Map declared fields marked @Id */
 
     private void mapComponents() {
-        parserData.forEachInjectedField(
-                (field, id, tag) -> tryMapComponentOrElement(field, id, tag));
+        parserData.forEachInjectedField(this::tryMapComponentOrElement);
     }
 
     private void tryMapComponentOrElement(Field field, String id, String tag) {

--- a/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java
+++ b/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/TemplateInitializer.java
@@ -25,7 +25,8 @@ import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
-import java.util.function.Function;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Stream;
@@ -62,18 +63,14 @@ public class TemplateInitializer {
     private static final Pattern TWO_WAY_BINDING_PATTERN = Pattern
             .compile("\\s*\\{\\{([^}:]*)(::[^}]*)?\\}\\}\\s*");
 
-    private static final ReflectionCache<PolymerTemplate<?>, ParserData> CACHE = new ReflectionCache<>(
-            clazz -> new ParserData());
+    private static final ConcurrentHashMap<TemplateParser, ReflectionCache<PolymerTemplate<?>, ParserData>> CACHE = new ConcurrentHashMap<>();
 
     private final PolymerTemplate<?> template;
     private final Class<? extends PolymerTemplate<?>> templateClass;
 
-    private final Map<String, Element> registeredElementIdToInjected = new HashMap<>();
-    private final boolean useCache;
     private final ParserData parserData;
-    private final Set<String> notInjectableElementIds = new HashSet<>();
 
-    private org.jsoup.nodes.Element parsedTemplateRoot;
+    private final Map<String, Element> registeredElementIdToInjected = new HashMap<>();
 
     private static class SubTemplateData {
         private final String id;
@@ -87,23 +84,86 @@ public class TemplateInitializer {
         }
     }
 
-    private static class ParserData
-            implements Function<String, Optional<String>> {
+    private interface InjectableFieldCunsumer {
+        public void apply(Field field, String id, String tag);
+    }
+
+    /**
+     * Immutable parser data which may be stored in cache.
+     */
+    private static class ParserData {
+
+        private final Map<String, String> tagById;
+        private final Map<Field, String> idByField;
+
+        private final Set<String> twoWayBindingPaths;
+
+        private final Collection<SubTemplateData> subTemplates;
+
+        private ParserData(Map<Field, String> fields, Map<String, String> tags,
+                Set<String> twoWayBindings,
+                Collection<SubTemplateData> subTemplates) {
+            tagById = Collections.unmodifiableMap(tags);
+            idByField = Collections.unmodifiableMap(fields);
+            twoWayBindingPaths = Collections.unmodifiableSet(twoWayBindings);
+            this.subTemplates = Collections
+                    .unmodifiableCollection(subTemplates);
+        }
+
+        private void forEachInjectedField(InjectableFieldCunsumer consumer) {
+            idByField.forEach(
+                    (field, id) -> consumer.apply(field, id, tagById.get(id)));
+        }
+
+        private Set<String> getTwoWayBindingPaths() {
+            return twoWayBindingPaths;
+        }
+
+        private void forEachSubTemplate(
+                Consumer<SubTemplateData> dataConsumer) {
+            subTemplates.forEach(dataConsumer);
+        }
+    }
+
+    /**
+     * Mutable temporary parser data which is used to collected data only during
+     * initialization. The data will be discarded and not stored anywhere.
+     */
+    private static class ParserDataContext {
         private final Map<String, String> tagById = new HashMap<>();
+        private final Map<Field, String> idByField = new HashMap<>();
+
         private final Collection<SubTemplateData> subTemplates = new ArrayList<>();
         private final Set<String> twoWayBindingPaths = new HashSet<>();
+        private final Set<String> notInjectableElementIds = new HashSet<>();
 
-        @Override
-        public Optional<String> apply(String id) {
-            return Optional.ofNullable(tagById.get(id));
-        }
-
-        private void addTag(String id, String tag) {
-            tagById.put(id, tag);
-        }
+        private org.jsoup.nodes.Element templateRoot;
 
         private void addSubTemplate(String id, String tag, JsonArray path) {
             subTemplates.add(new SubTemplateData(id, tag, path));
+        }
+
+        private void addNotInjectableId(String id) {
+            notInjectableElementIds.add(id);
+        }
+
+        private void addTwoWayBindingPath(String path) {
+            twoWayBindingPaths.add(path);
+        }
+
+        private void setTemplateRoot(org.jsoup.nodes.Element root) {
+            templateRoot = root;
+        }
+
+        private Optional<String> addTagName(String id, Field field) {
+            idByField.put(field, id);
+            Optional<String> tag = Optional
+                    .ofNullable(templateRoot.getElementById(id))
+                    .map(org.jsoup.nodes.Element::tagName);
+            if (tag.isPresent()) {
+                tagById.put(id, tag.get());
+            }
+            return tag;
         }
 
     }
@@ -120,20 +180,49 @@ public class TemplateInitializer {
     public TemplateInitializer(PolymerTemplate<?> template,
             TemplateParser parser) {
         this.template = template;
-        templateClass = (Class<? extends PolymerTemplate<?>>) template
-                .getClass();
 
         boolean productionMode = VaadinService.getCurrent()
                 .getDeploymentConfiguration().isProductionMode();
-        useCache = productionMode && CACHE.contains(templateClass);
-        parserData = productionMode ? CACHE.get(templateClass)
-                : new ParserData();
 
-        if (!useCache) {
-            parsedTemplateRoot = parser.getTemplateContent(templateClass,
-                    getElement().getTag());
-            parseTemplate();
+        templateClass = (Class<? extends PolymerTemplate<?>>) template
+                .getClass();
+
+        ParserData data = null;
+        if (productionMode) {
+            ReflectionCache<PolymerTemplate<?>, ParserData> cache = CACHE
+                    .computeIfAbsent(parser,
+                            analyzer -> new ReflectionCache<PolymerTemplate<?>, ParserData>(
+                                    clazz -> parseTemplate(clazz, analyzer)));
+            data = cache.get(templateClass);
         }
+        if (data == null) {
+            data = parseTemplate(templateClass, parser);
+        }
+        parserData = data;
+    }
+
+    private ParserData readData(ParserDataContext context) {
+        return new ParserData(context.idByField, context.tagById,
+                context.twoWayBindingPaths, context.subTemplates);
+    }
+
+    private ParserData parseTemplate(Class<? extends PolymerTemplate<?>> clazz,
+            TemplateParser parser) {
+        org.jsoup.nodes.Element root = parser.getTemplateContent(templateClass,
+                getElement().getTag());
+        ParserDataContext data = new ParserDataContext();
+        data.setTemplateRoot(root);
+        Elements templates = root.getElementsByTag("template");
+        for (org.jsoup.nodes.Element element : templates) {
+            org.jsoup.nodes.Element parent = element.parent();
+            if (parent != null && getElement().getTag().equals(parent.id())) {
+                inspectCustomElements(element, element, data);
+
+                inspectTwoWayBindings(element, data);
+            }
+        }
+        collectInjectedIds(clazz, data);
+        return readData(data);
     }
 
     /**
@@ -141,7 +230,7 @@ public class TemplateInitializer {
      */
     public void initChildElements() {
         registeredElementIdToInjected.clear();
-        mapComponents(templateClass);
+        mapComponents();
         createSubTemplates();
     }
 
@@ -151,41 +240,30 @@ public class TemplateInitializer {
      * @return an unmodifiable collection of two way binding paths
      */
     public Set<String> getTwoWayBindingPaths() {
-        return Collections.unmodifiableSet(parserData.twoWayBindingPaths);
+        return parserData.getTwoWayBindingPaths();
     }
 
     private void inspectCustomElements(org.jsoup.nodes.Element childElement,
-            org.jsoup.nodes.Element templateRoot) {
+            org.jsoup.nodes.Element templateRoot, ParserDataContext data) {
         if (isInsideTemplate(childElement, templateRoot)) {
-            storeNotInjectableElementId(childElement);
+            storeNotInjectableElementId(childElement, data);
         }
 
-        collectCustomElement(childElement, templateRoot);
-        childElement.children()
-                .forEach(child -> inspectCustomElements(child, templateRoot));
+        collectCustomElement(childElement, templateRoot, data);
+        childElement.children().forEach(
+                child -> inspectCustomElements(child, templateRoot, data));
     }
 
-    private void storeNotInjectableElementId(org.jsoup.nodes.Element element) {
+    private void storeNotInjectableElementId(org.jsoup.nodes.Element element,
+            ParserDataContext data) {
         String id = element.id();
         if (id != null && !id.isEmpty()) {
-            notInjectableElementIds.add(id);
+            data.addNotInjectableId(id);
         }
     }
 
-    private void parseTemplate() {
-        assert parsedTemplateRoot != null;
-        Elements templates = parsedTemplateRoot.getElementsByTag("template");
-        for (org.jsoup.nodes.Element element : templates) {
-            org.jsoup.nodes.Element parent = element.parent();
-            if (parent != null && getElement().getTag().equals(parent.id())) {
-                inspectCustomElements(element, element);
-
-                inspectTwoWayBindings(element);
-            }
-        }
-    }
-
-    private void inspectTwoWayBindings(org.jsoup.nodes.Element element) {
+    private void inspectTwoWayBindings(org.jsoup.nodes.Element element,
+            ParserDataContext data) {
         Matcher matcher = TWO_WAY_BINDING_PATTERN.matcher("");
         element.traverse(new NodeVisitor() {
             @Override
@@ -196,7 +274,7 @@ public class TemplateInitializer {
                     matcher.reset(attribute.getValue());
                     if (matcher.matches()) {
                         String path = matcher.group(1);
-                        parserData.twoWayBindingPaths.add(path);
+                        data.addTwoWayBindingPath(path);
                     }
                 }
             }
@@ -220,7 +298,7 @@ public class TemplateInitializer {
     }
 
     private void collectCustomElement(org.jsoup.nodes.Element element,
-            org.jsoup.nodes.Element templateRoot) {
+            org.jsoup.nodes.Element templateRoot, ParserDataContext data) {
         String tag = element.tagName();
 
         if (CustomElementRegistry.getInstance()
@@ -233,8 +311,7 @@ public class TemplateInitializer {
 
             String id = element.hasAttr("id") ? element.attr("id") : null;
             JsonArray path = getPath(element, templateRoot);
-            assert !useCache;
-            parserData.addSubTemplate(id, tag, path);
+            data.addSubTemplate(id, tag, path);
         }
     }
 
@@ -307,60 +384,52 @@ public class TemplateInitializer {
                 .orElseGet(() -> getElement().attachShadow());
     }
 
-    /* Map declared fields marked @Id */
-
-    private void mapComponents(Class<?> cls) {
+    private void collectInjectedIds(Class<?> cls, ParserDataContext data) {
         if (!AbstractTemplate.class.equals(cls.getSuperclass())) {
             // Parent fields
-            mapComponents(cls.getSuperclass());
+            collectInjectedIds(cls.getSuperclass(), data);
         }
 
         Stream.of(cls.getDeclaredFields()).filter(field -> !field.isSynthetic())
-                .forEach(field -> tryMapComponentOrElement(field));
+                .forEach(field -> collectedInjectedId(field, data));
     }
 
-    private void tryMapComponentOrElement(Field field) {
+    private void collectedInjectedId(Field field, ParserDataContext data) {
         Optional<Id> idAnnotation = AnnotationReader.getAnnotationFor(field,
                 Id.class);
         if (!idAnnotation.isPresent()) {
             return;
         }
         String id = idAnnotation.get().value();
-        if (notInjectableElementIds.contains(id)) {
+        if (data.notInjectableElementIds.contains(id)) {
             throw new IllegalStateException(String.format(
                     "Class '%s' contains field '%s' annotated with @Id('%s'). "
                             + "Corresponding element was found in a sub template, for which injection is not supported",
                     templateClass.getName(), field.getName(), id));
         }
 
-        Optional<String> tagName = getTagName(id);
-        if (!tagName.isPresent()) {
+        if (!data.addTagName(id, field).isPresent()) {
             throw new IllegalStateException(String.format(
                     "There is no element with "
                             + "id='%s' in the template file. Cannot map it using @%s",
                     id, Id.class.getSimpleName()));
         }
+    }
 
+    /* Map declared fields marked @Id */
+
+    private void mapComponents() {
+        parserData.forEachInjectedField(
+                (field, id, tag) -> tryMapComponentOrElement(field, id, tag));
+    }
+
+    private void tryMapComponentOrElement(Field field, String id, String tag) {
         Element element = getElementById(id).orElse(null);
 
         if (element == null) {
-            injectClientSideElement(tagName.get(), id, field);
+            injectClientSideElement(tag, id, field);
         } else {
             injectServerSideElement(element, field);
-        }
-    }
-
-    private Optional<String> getTagName(String id) {
-        if (useCache) {
-            return parserData.apply(id);
-        } else {
-            Optional<String> tag = Optional
-                    .ofNullable(parsedTemplateRoot.getElementById(id))
-                    .map(org.jsoup.nodes.Element::tagName);
-            if (tag.isPresent()) {
-                parserData.addTag(id, tag.get());
-            }
-            return tag;
         }
     }
 
@@ -471,9 +540,8 @@ public class TemplateInitializer {
     }
 
     private void createSubTemplates() {
-        parserData.subTemplates
-                .forEach(data -> doRequestAttachCustomElement(data.id, data.tag,
-                        data.path));
+        parserData.forEachSubTemplate(data -> doRequestAttachCustomElement(
+                data.id, data.tag, data.path));
     }
 
 }

--- a/flow-server/src/test/java/com/vaadin/flow/component/polymertemplate/DefaultTemplateParserTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/component/polymertemplate/DefaultTemplateParserTest.java
@@ -41,8 +41,6 @@ import org.mockito.Mockito;
 
 import com.vaadin.flow.component.Tag;
 import com.vaadin.flow.component.dependency.HtmlImport;
-import com.vaadin.flow.component.polymertemplate.DefaultTemplateParser;
-import com.vaadin.flow.component.polymertemplate.PolymerTemplate;
 import com.vaadin.flow.component.polymertemplate.PolymerTemplateTest.ModelClass;
 import com.vaadin.flow.internal.CurrentInstance;
 import com.vaadin.flow.server.DependencyFilter;
@@ -54,8 +52,8 @@ import com.vaadin.flow.server.VaadinUriResolverFactory;
 import com.vaadin.flow.server.WrappedHttpSession;
 import com.vaadin.flow.shared.VaadinUriResolver;
 import com.vaadin.flow.shared.ui.Dependency;
-import com.vaadin.flow.shared.ui.LoadMode;
 import com.vaadin.flow.shared.ui.Dependency.Type;
+import com.vaadin.flow.shared.ui.LoadMode;
 
 import net.jcip.annotations.NotThreadSafe;
 
@@ -153,8 +151,7 @@ public class DefaultTemplateParserTest {
 
     @Test
     public void defaultParser_hasTemplate_returnsContent() {
-        DefaultTemplateParser parser = new DefaultTemplateParser();
-        Element element = parser
+        Element element = DefaultTemplateParser.getInstance()
                 .getTemplateContent(ImportsInspectTemplate.class, "foo");
 
         Assert.assertTrue(element.getElementById("foo") != null);
@@ -169,8 +166,7 @@ public class DefaultTemplateParserTest {
                 new ByteArrayInputStream("<dom-module id='foo'></dom-module>"
                         .getBytes(StandardCharsets.UTF_8)));
 
-        DefaultTemplateParser parser = new DefaultTemplateParser();
-        Element element = parser
+        Element element = DefaultTemplateParser.getInstance()
                 .getTemplateContent(RootImportsInspectTemplate.class, "foo");
 
         Assert.assertTrue(element.getElementById("foo") != null);
@@ -183,8 +179,7 @@ public class DefaultTemplateParserTest {
 
         Mockito.when(request.getServletPath()).thenReturn("");
 
-        DefaultTemplateParser parser = new DefaultTemplateParser();
-        Element element = parser
+        Element element = DefaultTemplateParser.getInstance()
                 .getTemplateContent(ImportsInspectTemplate.class, "foo");
 
         Assert.assertTrue(element.getElementById("foo") != null);
@@ -210,8 +205,7 @@ public class DefaultTemplateParserTest {
                         "<dom-module id='foo'></dom-module>"
                                 .getBytes(StandardCharsets.UTF_8)));
 
-        DefaultTemplateParser parser = new DefaultTemplateParser();
-        Element element = parser
+        Element element = DefaultTemplateParser.getInstance()
                 .getTemplateContent(ImportsInspectTemplate.class, "foo");
 
         Assert.assertTrue(element.getElementById("foo") != null);
@@ -237,8 +231,7 @@ public class DefaultTemplateParserTest {
                         "<dom-module id='foo'></dom-module>"
                                 .getBytes(StandardCharsets.UTF_8)));
 
-        DefaultTemplateParser parser = new DefaultTemplateParser();
-        Element element = parser
+        Element element = DefaultTemplateParser.getInstance()
                 .getTemplateContent(ImportsInspectTemplate.class, "foo");
 
         Assert.assertTrue(element.getElementById("foo") != null);
@@ -256,8 +249,7 @@ public class DefaultTemplateParserTest {
                         "<!-- comment1 --><dom-module id='foo'><!-- comment2 --></dom-module>"
                                 .getBytes(StandardCharsets.UTF_8)));
 
-        DefaultTemplateParser parser = new DefaultTemplateParser();
-        Element element = parser
+        Element element = DefaultTemplateParser.getInstance()
                 .getTemplateContent(ImportsInspectTemplate.class, "foo");
         Assert.assertTrue(element.getElementById("foo") != null);
         assertThat("No comments should be present in the parsing result",
@@ -274,8 +266,8 @@ public class DefaultTemplateParserTest {
 
     @Test(expected = IllegalStateException.class)
     public void defaultParser_noTemplate_throws() {
-        DefaultTemplateParser parser = new DefaultTemplateParser();
-        parser.getTemplateContent(ImportsInspectTemplate.class, "bar1");
+        DefaultTemplateParser.getInstance()
+                .getTemplateContent(ImportsInspectTemplate.class, "bar1");
     }
 
     @Test(expected = IllegalStateException.class)
@@ -283,8 +275,8 @@ public class DefaultTemplateParserTest {
         Mockito.when(context.getResourceAsStream("/bar1.html"))
                 .thenReturn(null);
 
-        DefaultTemplateParser parser = new DefaultTemplateParser();
-        parser.getTemplateContent(ImportsInspectTemplate.class, "foo");
+        DefaultTemplateParser.getInstance()
+                .getTemplateContent(ImportsInspectTemplate.class, "foo");
     }
 
     @Test
@@ -301,7 +293,7 @@ public class DefaultTemplateParserTest {
         Mockito.when(service.getDependencyFilters())
                 .thenReturn(Collections.singletonList(filter));
 
-        DefaultTemplateParser parser = new DefaultTemplateParser();
+        TemplateParser parser = DefaultTemplateParser.getInstance();
 
         Element element = parser
                 .getTemplateContent(ImportsInspectTemplate.class, "foo");
@@ -331,8 +323,8 @@ public class DefaultTemplateParserTest {
         Mockito.when(service.getDependencyFilters())
                 .thenReturn(Collections.singletonList(filter));
 
-        DefaultTemplateParser parser = new DefaultTemplateParser();
-        parser.getTemplateContent(ImportsInspectTemplate.class, "foo");
+        DefaultTemplateParser.getInstance()
+                .getTemplateContent(ImportsInspectTemplate.class, "foo");
     }
 
     @Test(expected = IllegalStateException.class)
@@ -349,8 +341,8 @@ public class DefaultTemplateParserTest {
         Mockito.when(service.getDependencyFilters())
                 .thenReturn(Collections.singletonList(filter));
 
-        DefaultTemplateParser parser = new DefaultTemplateParser();
-        parser.getTemplateContent(ImportsInspectTemplate.class, "foo");
+        DefaultTemplateParser.getInstance()
+                .getTemplateContent(ImportsInspectTemplate.class, "foo");
     }
 
 }

--- a/flow-server/src/test/java/com/vaadin/flow/component/polymertemplate/TemplateInitializerTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/component/polymertemplate/TemplateInitializerTest.java
@@ -90,28 +90,25 @@ public class TemplateInitializerTest extends HasCurrentService {
         String outsideTemplateElementId = OutsideTemplateClass.class
                 .getField("element").getAnnotation(Id.class).value();
 
-        templateParser = (clazz,
-                tag) -> Jsoup.parse(String.format(
-                        "<dom-module id='%s'><template>"
-                                + "    <template><div id='%s'>Test</div></template>"
-                                + "    <div id='%s'></div>"
-                                + "    <div a='{{twoWay}}' b='{{invalid}} syntax' c='{{two.way}}'"
-                                + "        d='{{invalidSyntax' e='{{withEvent::eventName}}' f='[[oneWay]]'></div>"
-                                + "</template></dom-module>",
-                        parentTemplateId, inTemplateElementId,
-                        outsideTemplateElementId));
+        templateParser = (clazz, tag) -> Jsoup.parse(String.format(
+                "<dom-module id='%s'><template>"
+                        + "    <template><div id='%s'>Test</div></template>"
+                        + "    <div id='%s'></div>"
+                        + "    <div a='{{twoWay}}' b='{{invalid}} syntax' c='{{two.way}}'"
+                        + "        d='{{invalidSyntax' e='{{withEvent::eventName}}' f='[[oneWay]]'></div>"
+                        + "</template></dom-module>",
+                parentTemplateId, inTemplateElementId,
+                outsideTemplateElementId));
     }
 
     @Test(expected = IllegalStateException.class)
     public void inTemplateShouldThrowAnException() {
-        new TemplateInitializer(new InTemplateClass(), templateParser)
-                .initChildElements();
+        new TemplateInitializer(new InTemplateClass(), templateParser);
     }
 
     @Test
     public void outsideTemplateShouldNotThrowAnException() {
-        new TemplateInitializer(new OutsideTemplateClass(), templateParser)
-                .initChildElements();
+        new TemplateInitializer(new OutsideTemplateClass(), templateParser);
     }
 
     @Test

